### DISCUSSION TITLE
Help our team by reviewing an idea :)

### DISCUSSION THREAD

**Discussion by win-luck** [euphony-io/euphony#169] [Q&A]
E조 Soundary 팀의 팀장 채승운입니다.

우리 앱은 두 기기가 음파를 발생시키고, 서로를 인식하는 데 걸리는 시간을 계산하여, 두 사람의 사이의 거리를 확보합니다.
Euphony_hub_collector는 기기 자신이 입력받은 문자를 통해 음파를 발생시키고, 기기 자신이 그 음파를 다시 문자로 변환하여 Success를 출력합니다.

후자와 달리 우리 앱은 서로를 인식하면서도, 자기 스스로의 음파를 인식해서는 안 되므로,
한 기기가 다른 기기에 보내는 문자열을 "발생시간/문자열/ID"의 형태로 전송하는 것은 어떨까? 라는 아이디어를 떠올렸습니다.
각 기기의 listener들은 문자열을 받았을 때 뒤의 ID가 자신의 것인지 확인한 후 그렇지 않을 경우에만 문자열 속에 같이 속해있는 발생시간을 활용해 거리를 계산하는 과정을 진행할 것입니다. 거리는 4~5m 이내의 경우에만 화면에 dialog 형태로 출력할 것입니다.

이 아이디어 대해서 검토해주시면 감사하겠습니다:)
DB 등의 구축은 계획에 없고, 오직 두 기기의 상황만 프로토타입으로 구현해 보고자 합니다.
통신 개념에 대해 많이 미흡합니다. 친절한 설명해주시면 정말로 감사하겠습니다.

**Comment by designe** [comment]
@chaso1495 승운님 안녕하세요.
좋은 질문 감사드립니다 👍  

우선 구현 방안은 다양하겠지만(?) 제가 생각했을때 이것을 가장 쉽게 구현하는 방법은 EuPI를 사용하는 것입니다.
- speaker side : https://github.com/euphony-io/euphony-speaker/blob/development/app/src/main/java/co/jbear/euphony_speaker/fragments/EuPIFragment.java

아래는 java에서 사용하는 api 예시인데요.
```java
txManager = new EuTxManager(this.getContext());

 txManager.callEuPI(18500, EuTxManager.EuPIDuration.LENGTH_LONG); // 500ms
// txManager.callEuPI(18500, EuTxManager.EuPIDuration.LENGTH_SHORT); // 200ms
// txManager.callEuPI(18500, EuTxManager.EuPIDuration.LENGTH_FOREVER); // FOREVER
```

`callEuPI`를 사용하시면 앞의 해당 주파수 18500이든 원하시는 주파수를 넣으시면 되구요.
이를 FOREVER하게 발생시키면 18500 주파수를 계속 발생시키게 됩니다. 
그럼 listener로 해당 주파수를 인식하는 코드를 넣으면 이를 인식하게 되는데요. 아래가 예시 fragment입니다.

- listener side : https://github.com/euphony-io/euphony-listener/blob/development/app/src/main/java/co/jbear/euphony_listener/fragments/EuPICallFragment.java

```java
int api_1_count = 0;
int api_2_count = 0;
int api_3_count = 0;

mRxManager = new EuRxManager(EuOption.ModeType.EUPI);

mRxManager.setOnWaveKeyPressed(18500, new EuPICallDetector() {
    @Override
    public void call() {
        api_1_count++;
        API_1_COUNT.setText("" + api_1_count);
    }
});

mRxManager.setOnWaveKeyDown(19000, new EuPICallDetector() {
    @Override
    public void call() {
        api_2_count++;
        API_2_COUNT.setText("" + api_2_count);
    }
});

mRxManager.setOnWaveKeyUp(20000, new EuPICallDetector() {
    @Override
    public void call() {
        api_3_count++;
        API_3_COUNT.setText("" + api_3_count);
    }
});
```
위 코드는 `setOnWaveKeyPressed`, `setOnWaveKeyDown`, `setOnWaveKeyUp` 함수 3개를 사용하는데요. 
 `setOnWaveKeyPressed` : 해당 음파(argument 첫번째 값)이 인식될때 계속 호출되는 API
 `setOnWaveKeyDown` : 이거는 음파가 발생한 최초 1회 인식하는 API
 `setOnWaveKeyUp` : 음파가 발생하고 꺼졌을때 최초 1회 인식하는 API

이걸 참고하시면 해당 아이디어를 구현하실때 도움이 될 것 같아요.

각 기기마다 고유의 주파수를 지정해서 발생시키고 이를 다른 기기에서 주파수가 잡히는지를 확인하면 우선은 간단하게 prototyping 구현이 될 것 같아요.
하지만 아쉽게도 EuPI는 문서화가 되어 있지 않아요 😢 
혹시라도 쓰시다가 문서화 기여도 해주시면 좋겠네요!! 😂 

그리고 EuPI는 jcenter artifact를 사용하는 것이 아닌 euphony를 직접 빌드해서 나온 aar 파일을 import 해서 쓰셔야 합니다.
최신 euphony repo를 clone 해서 빌드하신 후 추출된 aar을 프로젝트에 import하시면 됩니다.
aar file import 방법은 아래 참고하시면 되겠습니다!

https://stackoverflow.com/questions/16682847/how-to-manually-include-external-aar-package-using-gradle-for-android

**Reply by win-luck** [reply]
![image](https://user-images.githubusercontent.com/53044069/180996525-9d138506-feca-4c53-990e-ba56539c43a6.png)

함수가 아예 없다고 뜨는데 혹시 제가 뭘 잘못한걸까요..?

**Reply by designe** [reply]
euphony 최신 버전을 aar로 추출하셨나요?
https://github.com/euphony-io/euphony/blob/73ed9f09118847e93236482e38bdbe47e7d15ff5/euphony/src/main/java/co/euphony/rx/EuRxManager.java#L102

함수가 아예 안뜨는 경우에는 다른 버전(?)을 사용하고 계신 것 같은데
aar 추출 및 라이브러리 포팅을 어떻게 하셔서 쓰는지 설명을 써주시면 해결책을 찾는데 도움이 될 것 같아요.

**Reply by win-luck** [reply]
![image](https://user-images.githubusercontent.com/53044069/181218179-54cdb500-c5fe-42f6-88d5-e33d83bc21ff.png)
![image](https://user-images.githubusercontent.com/53044069/181218331-a1ed9f28-3888-44df-b50a-968f0b44b19a.png)
![image](https://user-images.githubusercontent.com/53044069/181231005-daf748c9-fabf-40bb-9dc4-789ab6b4de0c.png)



최신 버전 추출 시도 중인데 자꾸 에러가 뜨네요 혹시 뭐가 문제인지 알 수 있을까요? 빌드가 되어야 aar 파일을 얻어서 본 프로젝트에 사용할 수 있을 텐데 걱정이네요 ㅜ

**Reply by designe** [reply]
https://github.com/euphony-io/euphony/discussions/146#discussioncomment-3147907

요기참조하시믄 됩니당!!

**Reply by win-luck** [reply]
![image](https://user-images.githubusercontent.com/53044069/181229794-10ab508d-2323-4191-aeb9-68d025b996ce.png)
![image](https://user-images.githubusercontent.com/53044069/181230179-8c1d6294-228c-432a-8a29-7e0b84284a31.png)

말씀하신 대로 대처했는데 또 난리네요,,

**Reply by designe** [reply]
@chaso1495 빌드 실패 로그 카피해서 여기 붙여주시면 확인해볼 수 있을거 같아요. 이미지로는 상세한 에러 로그 확인이 안되네요~!

build clean
rebuild 했을때 로그 주시면 좋을 것 같아요~!
